test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageWriterTests.cs

```diff
@@ -487,6 +491,361 @@ public async Task WhenInjectingStreamBasedJsonWriterFactoryAsync_CreatesWriterUs
+            personType.AddKeys(
+                personType.AddStructuralProperty("Id", EdmPrimitiveTypeKind.Int32));
+            personType.AddStructuralProperty("Name", EdmPrimitiveTypeKind.String);
+            personType.AddStructuralProperty(
+                "Emails",
+                new EdmCollectionTypeReference(
+                    new EdmCollectionType(EdmCoreModel.Instance.GetString(isNullable: false))));
+            EdmComplexType addressType = model.AddComplexType("ns", "Address");
+            addressType.AddStructuralProperty("City", EdmPrimitiveTypeKind.String);
+            addressType.AddStructuralProperty("Country", EdmPrimitiveTypeKind.String);
+            personType.AddStructuralProperty("Address", new EdmComplexTypeReference(addressType, isNullable: false));
+            IEdmEntitySet peopleSet = model.AddEntityContainer("ns", "Service").AddEntitySet("People", personType);
+
+
+            string source = "{" +
+                @"""Id"": 1," +
+                @"""Name"": ""John""," +
+                @"""Emails"":[""john@mailer.com"",""john@work.com""]," +
+                @"""Address"":{""City"":""Nairobi"",""Country"":""Kenya""}" +
+                @"}";
+            var json = JsonDocument.Parse(source);
+            var jsonRoot = json.RootElement;
+            var resource = new ODataResource
+            {
+                Properties = new[]
+                {
+                    new ODataProperty { Name = "Id", Value = new ODataJsonElementValue(jsonRoot.GetProperty("Id")) },
+                    new ODataProperty { Name = "Name", Value = new ODataJsonElementValue(jsonRoot.GetProperty("Name")) },
+                    new ODataProperty { Name = "Emails", Value = new ODataJsonElementValue(jsonRoot.GetProperty("Emails")) },
+                    new ODataProperty { Name = "Address", Value = new ODataJsonElementValue(jsonRoot.GetProperty("Address")) }
+                }
+            };
+
+            // Act
+            string result = WriteAndGetPayload(
+                model,
+                "application/json; charset=utf-8",
+                writer =>
+                {
+                    var resourceWriter = writer.CreateODataResourceWriter(peopleSet);
+                    resourceWriter.WriteStart(resource);
+                    resourceWriter.WriteEnd();
+                },
+                path: new ODataPath(new EntitySetSegment(peopleSet)));
+
+            // Result
+            string expected = @"{""@odata.context"":""http://www.example.com/$metadata#People/$entity"",""Id"":1,""Name"":""John"",""Emails"":[""john@mailer.com"",""john@work.com""],""Address"":{""City"":""Nairobi"",""Country"":""Kenya""}}";
+            Assert.Equal(expected, result);
+        }
+
+        [Fact]
+        public async Task WriteEntityWithJsonElementValues_Async()
+        {
+            // Arrange
+            EdmModel model = new EdmModel();

```

Comment:
The model creation codes repeats? could be merge them as a method?

Reply:
Yeah. But I create different model elements based on scenario. I thought of merging them into a shared method, but since this test file is large, I thought things could get lost if I created a helper method that's only used in 3 tests.